src/libponyc/pass/verify.c

```diff
@@ -90,19 +96,386 @@ static bool verify_assign_lvalue(pass_opt_t* opt, ast_t* ast)
+    case TK_LETREF:
+    case TK_VARREF:
+    {
+      temp_ast = ast_child(temp_ast);
+      len += strlen(ast_name(temp_ast));
+      break;
+    }
+
+    case TK_THIS:
+    {
+      temp_ast = ast_sibling(temp_ast);
+      // string len already added by loop above
+      // subtract 1 because we don't have to add '.'
+      // since we're ignoring the `this`
+      len -= 1;
+      break;
+    }
+
+    default:
+    {
+      pony_assert(0);
+    }
+  }
+
+  // for the \0 at the end
+  len = len + 1;
+
+  char* buf = (char*)ponyint_pool_alloc_size(len);
+  size_t offset = 0;
+  const char* name = ast_name(temp_ast);
+  size_t slen = strlen(name);
+  strncpy(buf + offset, name, slen);
+  offset += slen;
+  temp_ast = ast_parent(temp_ast);
+
+  while(temp_ast != ast)
+  {
+    buf[offset] = '.';
+    offset += 1;
+    temp_ast = ast_sibling(temp_ast);
+    name = ast_name(temp_ast);
+    slen = strlen(name);
+    strncpy(buf + offset, name, slen);
+    offset += slen;
+    temp_ast = ast_parent(temp_ast);
+  }
+
+  pony_assert((offset + 1) == len);
+  buf[offset] = '\0';
+
+  return stringtab_consume(buf, len);
+}
+
+static ast_t* get_fun_def(pass_opt_t* opt, ast_t* ast)
+{
+  pony_assert((ast_id(ast) == TK_FUNREF) || (ast_id(ast) == TK_FUNCHAIN) ||
+    (ast_id(ast) == TK_NEWREF));
+  AST_GET_CHILDREN(ast, receiver, method);
+
+  // Receiver might be wrapped in another funref/newref
+  // if the method had type parameters for qualification.
+  if(ast_id(receiver) == ast_id(ast))
+    AST_GET_CHILDREN_NO_DECL(receiver, receiver, method);
+
+  // Look up the original method definition for this method call.
+  deferred_reification_t* method_def = lookup_try(opt, receiver,
+    ast_type(receiver), ast_name(method), true);
+
+  pony_assert(method_def != NULL);
+
+  ast_t* method_ast = method_def->ast;
+  deferred_reify_free(method_def);
+
+  pony_assert(ast_id(method_ast) == TK_FUN || ast_id(method_ast) == TK_BE ||
+    ast_id(method_ast) == TK_NEW);
+
+  // behavior call
+  if(ast_id(method_ast) == TK_BE)
+    return NULL;
+  else
+    return method_ast;
+}
+
+static bool verify_fun_field_not_referenced(pass_opt_t* opt, ast_t* ast,
+  ast_t* consume_ast, const char* origin_type_name, ast_t* consumed_type,
+  const char* consumed_field_name, const char* consumed_field_full_name,
+  consume_funs_t* consume_funs_map)
+{
+  token_id ast_token = ast_id(ast);
+  if((ast_token == TK_FUNREF) || (ast_token == TK_NEWREF)
+    || (ast_token == TK_FUNCHAIN))
+  {
+    ast_t* fun_def = get_fun_def(opt, ast);
+
+    if(fun_def != NULL)
+    {
+      size_t index = HASHMAP_UNKNOWN;
+
+      ast_t** fref = consume_funs_get(consume_funs_map, &fun_def, &index);
+
+      // add to hashmap and evaluate if we haven't seen this function yet
+      if(fref == NULL)
+      {
+        // add to hashmap
+        consume_funs_putindex(consume_funs_map, &fun_def, index);
+
+        // recursive check child
+        // TODO: open question, print full function call tree?
```

Comment:
Personally, I don't think it's necessary to print the whole tree - but maybe someone will chime in with a request for this in the future and we could discuss further at that time.